API 1 of 3: Refactor Organisation Model

R2ZER0 opened the pull request:
This PR was originally Part 1 of https://github.com/ThreeSixtyGiving/datastore/pull/175

This PR:
* Refactors out a lot of the Organisation-fetching logic from the OrganisationDetailView, to the Organisation "model" (dataclass) itsself, so it can be reused.
* Hides `Publisher.getter_run` from the public API

R2ZER0 committed:
api: Add Organisation.get method

Move the logic to create an Organisation "fake" model instance from db, out of
the view and into the model itsself.

R2ZER0 committed:
api: Add drf-spectacular to requirements & settings

drf-spectacular is a library which extends rest framework to generate OpenAPI
Schema 3.0, and optionally server a live docs site based on the schema.

R2ZER0 committed:
api: Fix KeyError when looking up grants for an unknown Org

TODO: Perhaps nn future, explicitly handle and return 404, rather than empty list?

R2ZER0 committed:
api: Don't expose Publisher.getter_run in API

In preparation for removing this field from the Model.

R2ZER0 committed:
api: Rename Org Serialisers due to naming conflict

Names conflict with dashboard serialisers, drf-spectacular requires serialisers
have unique names.

R2ZER0 committed:
api: Annotate Org API w/ OpenAPI types

R2ZER0 committed:
api: Add schema & docs site URLs

From drf-spectacular

R2ZER0 committed:
api: Update drf-spectacular

R2ZER0 committed:
api: Remove GrantDetailView

To simplify the API, remove redundant single grant detail view.
The other views display all the same data.

R2ZER0 committed:
api: Add name to top-level Org model

R2ZER0 committed:
api: Serializer for Funder/Recipient aggregate data

R2ZER0 committed:
api: Update README with how to update OpenAPI schema

R2ZER0 committed:
api: Remove unused import

R2ZER0 committed:
api: Add 360G Schema converted to OpenAPI

and restrict the schema/swagger-ui to display only public endpoints.

R2ZER0 committed:
api: Refactor, into own module

* Move the new Org API into it's own API submodule
* Make use of dataclass serializer library

R2ZER0 committed:
api: Remove unused default schema view

R2ZER0 committed:
tests: Add tests for Org API

michaelwood commented on datastore/api/experimental/api.py:
```diff
@@ -112,7 +97,7 @@ class OrganisationGrantsMadeView(generics.ListAPIView):
 
     def get_queryset(self):
-        org_id = self.kwargs["org_id"]
+        org_id = self.kwargs.get("org_id")
```
Maybe a quick comment about why you want an empty list if the org_id is None

michaelwood commented on datastore/api/experimental/api.py:
```diff
@@ -74,36 +74,21 @@ def get_object(self):
-            ).order_by("-getter_run__datetime")[0]
-        except IndexError:
-            publisher = None
+        organisation = api_experimental.Organisation.get(
```
At some point we'll need to change this from `experimental`

R2ZER0 committed:
api: Raise 404 on GrantsView if Org doesn't exist

R2ZER0 committed:
tests: Add test for Organisation.exists

R2ZER0 committed:
tests: Fix org_api recipient grants test

Fix a bug in this test missed due toprevious GrantsMadeView behaviour of returning empty list instead of 404.

R2ZER0 committed:
tests: Add Org API tests for non-existent Org behaviour

R2ZER0 committed:
api: Fix bug in OrgDetail for non-existent Org

The OrgDetail view wasn't properly handling object-not-found errors.

R2ZER0 committed:
tests: Fix example OrgIDs to be valid OrgIDs

Example OrgIDs should start with prefix XE-EXAMPLE

R2ZER0 committed:
api: Use absolute imports

Replace relative imports with absolute imports

R2ZER0 committed:
tests: Add tests for Organisation.get()
